rclpy/rclpy/context.py

```diff
@@ -28,11 +28,32 @@ def __init__(self):
         from rclpy.impl.implementation_singleton import rclpy_implementation
         self._handle = rclpy_implementation.rclpy_create_context()
         self._lock = threading.Lock()
+        self._guard_lock = threading.RLock()
+        self._guard_conditions = {}
```

Comment:
I recommend against having the `Context` store guard conditions. It seems too specific to one task: waking the executor. Neither wait sets nor guard conditions are things the `Context` needs to know about.

Instead I recommend adding APIs to register a callback to call when the `Context` is shut down. Something like [`atexit.register()` and `atexit.unregister()`](https://docs.python.org/3/library/atexit.html), but on a `Context` instance.

That would allow an executor to register a callback that calls `executor.wake()`. A small benefit is it would reuse the existing guard condition rather than create a second one.

Reply:
I've applied this changes in https://github.com/ros2/rclpy/pull/403/commits/2918d3480ade7f84bf838cf370a787623cb53343.